src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py

```diff
@@ -0,0 +1,233 @@
+    f'--timeout={_TIMEOUT}',
+    f'--fork_server={_SERVER_COUNT}',
+    f'--rss_limit_mb={_RSS_LIMIT}',
+    f'--rlimit_as_mb={_RLIMIT_AS}',
+    f'--address_space_limit_mb={_ADDRESS_SPACE_LIMIT}',
+]
+
+_CRASH_REGEX = re.compile(
+    'centipede.cc:572] ReportCrash[0]: crash detected, saving input to (.*)')
+
+
+class CentipedeError(Exception):
+  """Base exception class."""
+
+
+def _get_runner():
+  """Get the Centipede runner."""
+  centipede_path = os.path.join(environment.get_value('BUILD_DIR'), 'centipede')
+  if not os.path.exists(centipede_path):
+    raise CentipedeError('Centipede not found in build')
+
+  os.chmod(centipede_path, 0o755)
+  if environment.get_value('USE_UNSHARE'):
+    return new_process.UnicodeModifierRunner(centipede_path)
+
+  return new_process.UnicodeProcessRunner(centipede_path)
+
+
+def _get_reproducer_path(line):
+  """Get the reproducer path, if any."""
+  crash_match = _CRASH_REGEX.match(line)
+  if not crash_match:
+    return None
+
+  return crash_match.group(1)
+
+
+class Engine(engine.Engine):
+  """Centipede engine implementation."""
+
+  @property
+  def name(self):
+    return 'centipede'
+
+  def prepare(self, corpus_dir, target_path, build_dir):  # pylint: disable=unused-argument
+    """Prepare for a fuzzing session, by generating options. Returns a
+    FuzzOptions object.
+
+    Args:
+      corpus_dir: The main corpus directory.
+      target_path: Path to the target.
+      build_dir: Path to the build directory.
+
+    Returns:
+      A FuzzOptions object.
+    """
+    os.chmod(target_path, 0o775)
+    arguments = []
+    dict_path = dictionary_manager.get_default_dictionary_path(target_path)
+    if os.path.exists(dict_path):
+      arguments.append(f'--dictionary={dict_path}')
+
+    # Is it OK to create workdir with this function?
+    # workdir saves centipede-readable corpus&feature files, and crashes
+    workdir = self._create_temp_dir('workdir')
+    # Will the workdir always exist?
+    arguments.append(f'--workdir={workdir}')
+
+    # Will the corpus directory always exist?
+    Path(corpus_dir).mkdir(exist_ok=True)
+    arguments.append(f'--corpus_dir={corpus_dir}')
+
+    # Download sanitized binaries.
+    sanitized_binaries = glob.glob(
+        os.path.join(build_dir, '*_*', os.path.basename(target_path)))
+    arguments.append(f'--extra_binaries={",".join(sanitized_binaries)}')
+
+    return engine.FuzzOptions(corpus_dir, arguments, {})
+
+  def fuzz(self, target_path, options, reproducers_dir, max_time):  # pylint: disable=unused-argument
+    """Run a fuzz session.
+
+    Args:
+      target_path: Path to the target.
+      options: The FuzzOptions object returned by prepare().
+      reproducers_dir: The directory to put reproducers in when crashes
+          are found.
+      max_time: Maximum allowed time for the fuzzing to run.
+
+   Returns:
+      A FuzzResult object.
+    """
+    runner = _get_runner()
+    arguments = _DEFAULT_ARGUMENTS[:]
```

Comment:
probably better to do _DEFAULT_ARGUMENTS.copy() as it's less weird (even though [:] is well known in python, see https://stackoverflow.com/questions/2612802/how-do-i-clone-a-list-so-that-it-doesnt-change-unexpectedly-after-assignment)